src/metabase/sync/analyze/query_results.clj

```diff
@@ -0,0 +1,84 @@
+(ns metabase.sync.analyze.query-results
+  "Analysis similar to what we do as part of the Sync process, but aimed at analyzing and introspecting query
+  results. The current focus of this namespace is around column metadata from the results of a query. Going forward
+  this is likely to extend beyond just metadata about columns but also about the query results as a whole and over
+  time."
+  (:require [metabase.models.humanization :as humanization]
+            [metabase.query-processor.interface :as i]
+            [metabase.sync.analyze.classifiers.name :as classify-name]
+            [metabase.sync.analyze.fingerprint :as f]
+            [metabase.sync.interface :as si]
+            [metabase.util :as u]
+            [metabase.util.schema :as su]
+            [schema.core :as s]))
+
+(def ^:private DateTimeUnitKeywordOrString
+  "Schema for a valid datetime unit string like \"default\" or \"minute-of-hour\"."
+  (s/constrained su/KeywordOrString
+                 (fn [unit]
+                   (contains? i/datetime-field-units (keyword unit)))
+                 "Valid field datetime unit keyword or string"))
+
+(def ^:private ResultColumnMetadata
+  "Result metadata for a single column"
+  {:name                          su/NonBlankString
+   :display_name                  su/NonBlankString
+   (s/optional-key :description)  (s/maybe su/NonBlankString)
+   :base_type                     su/FieldTypeKeywordOrString
+   (s/optional-key :special_type) (s/maybe su/FieldTypeKeywordOrString)
+   (s/optional-key :unit)         (s/maybe DateTimeUnitKeywordOrString)
+   (s/optional-key :fingerprint)  (s/maybe si/Fingerprint)})
+
+(def ResultsMetadata
+  "Schema for valid values of the `result_metadata` column."
+  (su/with-api-error-message (s/named [ResultColumnMetadata]
+                                      "Valid array of results column metadata maps")
+    "value must be an array of valid results column metadata maps."))
+
+(s/defn ^:private maybe-infer-special-type :- ResultColumnMetadata
+  "If we already have a special type from the stored metadata, use that, otherwise compute the special type similar to
+  how our sync process would compute it for a new field."
+  [result-metadata col]
+  (if (contains? result-metadata :special_type)
+    result-metadata
+    (assoc result-metadata :special_type (classify-name/infer-special-type col))))
```

Comment:
infer-special-type already has rules when to rerun. I think it's better to defer to it completely.

Reply:
Do you mean there's no need to run it here? What about in the case of native queries where it wouldn't have ran at all?

Reply:
Sorry I was unclear. I meant just the check. The whole fn can collapse into just 
```
(assoc result-metadata :special_type (classify-name/infer-special-type col))
```